Older macOS do not provide `vdprintf`: `Undefined symbols: "_vdprintf"`

@adrienverge The source uses `vdprintf` just in a single instance. Is it possible to add a fallback?

```
  CC       src/openfortivpn-userinput.o
src/userinput.c: In function ‘pinentry_exchange’:
src/userinput.c:176: warning: implicit declaration of function ‘vdprintf’
  CCLD     openfortivpn
Undefined symbols:
  "_vdprintf", referenced from:
      _pinentry_exchange in openfortivpn-userinput.o
ld: symbol(s) not found
```

Yeah, looks good:
```
--->  Configuring openfortivpn
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_openfortivpn/openfortivpn/work/openfortivpn-1.22.0" && autoreconf --install --verbose 
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: /opt/local/bin/aclocal 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /opt/local/bin/autoconf
autoreconf: configure.ac: not using Autoheader
autoreconf: running: /opt/local/bin/automake --add-missing --copy --no-force
configure.ac:10: installing './compile'
configure.ac:7: installing './install-sh'
configure.ac:7: installing './missing'
Makefile.am: installing './depcomp'
autoreconf: Leaving directory '.'
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_openfortivpn/openfortivpn/work/openfortivpn-1.22.0" && ./configure --prefix=/opt/local 
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... /opt/local/bin/gmkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... /usr/bin/gcc-4.2
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether /usr/bin/gcc-4.2 accepts -g... yes
checking for /usr/bin/gcc-4.2 option to enable C11 features... unsupported
checking for /usr/bin/gcc-4.2 option to enable C99 features... -std=gnu99
checking whether /usr/bin/gcc-4.2 -std=gnu99 understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of /usr/bin/gcc-4.2 -std=gnu99... gcc3
checking for a sed that does not truncate output... /opt/local/bin/gsed
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for wchar.h... yes
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether _XOPEN_SOURCE should be defined... no
checking for pkg-config... /opt/local/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking whether make supports nested variables... (cached) yes
checking for libssl >= 1.0.2 libcrypto >= 1.0.2... yes
checking for pthread_create in -lpthread... yes
checking for forkpty in -lutil... yes
checking for libsystemd... no
libsystemd not present
checking for arpa/inet.h... yes
checking for fcntl.h... yes
checking for getopt.h... yes
checking for ifaddrs.h... yes
checking for netdb.h... yes
checking for netinet/in.h... yes
checking for netinet/tcp.h... yes
checking for pthread.h... yes
checking for strings.h... (cached) yes
checking for sys/ioctl.h... yes
checking for syslog.h... yes
checking for sys/socket.h... yes
checking for sys/stat.h... (cached) yes
checking for sys/types.h... (cached) yes
checking for sys/wait.h... yes
checking for termios.h... yes
checking for unistd.h... (cached) yes
checking for net/if.h... yes
checking for net/route.h... yes
checking for libutil.h... no
checking for mach/mach.h... yes
checking for pty.h... no
checking for semaphore.h... yes
checking for util.h... yes
checking for net/if_var.h... yes
checking for an ANSI C-conforming const... yes
checking for inline... inline
checking for working volatile... yes
checking for off_t... yes
checking for pid_t... yes
checking for size_t... yes
checking for ssize_t... yes
checking for uint16_t... yes
checking for uint32_t... yes
checking for uint8_t... yes
checking for struct termios... yes
checking for access... yes
checking for close... yes
checking for connect... yes
checking for execv... yes
checking for _exit... yes
checking for fcntl... yes
checking for fileno... yes
checking for forkpty... yes
checking for freeaddrinfo... yes
checking for freeifaddrs... yes
checking for gai_strerror... yes
checking for getaddrinfo... yes
checking for geteuid... yes
checking for getifaddrs... yes
checking for getopt_long... yes
checking for htons... yes
checking for inet_addr... yes
checking for inet_ntoa... yes
checking for ioctl... yes
checking for isatty... yes
checking for memmem... yes
checking for ntohs... yes
checking for open... yes
checking for openlog... yes
checking for pclose... yes
checking for popen... yes
checking for pthread_cancel... yes
checking for pthread_cond_init... yes
checking for pthread_cond_signal... yes
checking for pthread_cond_wait... yes
checking for pthread_create... yes
checking for pthread_join... yes
checking for pthread_mutexattr_init... yes
checking for pthread_mutex_destroy... yes
checking for pthread_mutex_init... yes
checking for pthread_mutex_lock... yes
checking for pthread_mutex_unlock... yes
checking for pthread_self... yes
checking for pthread_sigmask... yes
checking for read... yes
checking for select... yes
checking for sem_destroy... yes
checking for sem_init... yes
checking for sem_post... yes
checking for sem_wait... yes
checking for setenv... yes
checking for setsockopt... yes
checking for sigaddset... yes
checking for sigemptyset... yes
checking for sleep... yes
checking for socket... yes
checking for strcasecmp... yes
checking for strdup... yes
checking for strncasecmp... yes
checking for strsignal... yes
checking for strtok_r... yes
checking for syslog... yes
checking for tcgetattr... yes
checking for tcsetattr... yes
checking for usleep... yes
checking for vsyslog... yes
checking for waitpid... yes
checking for write... yes
checking for pthread_mutexattr_setrobust... no
checking for vdprintf... no
checking whether rtentry is available and has rt_dst... no
checking for /proc/net/route... no
checking for /usr/sbin/netstat... yes
checking for ppp... /usr/sbin/ppp
checking for /usr/sbin/ppp... no
checking for pppd... /usr/sbin/pppd
checking for /usr/sbin/pppd... yes
configure: HAVE_USR_SBIN_PPP... 0
configure: HAVE_USR_SBIN_PPPD... 1
configure: LEGACY_PPPD... 0
configure: HAVE_PROC_NET_ROUTE... 0
configure: PPP_PATH... /usr/sbin/pppd
configure: NETSTAT_PATH... /usr/sbin/netstat
checking for resolvconf... DISABLED
configure: RESOLVCONF_PATH... DISABLED
configure: HAVE_RESOLVCONF... 0
configure: USE_RESOLVCONF... 0
Package systemd was not found in the pkg-config search path.
Perhaps you should add the directory containing `systemd.pc'
to the PKG_CONFIG_PATH environment variable
No package 'systemd' found
configure: systemdsystemunitdir... 
configure: HAVE_SO_BINDTODEVICE... 0
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: executing depfiles commands
config.status: executing timestamp commands
--->  Building openfortivpn
Executing:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_openfortivpn/openfortivpn/work/openfortivpn-1.22.0" && /usr/bin/make -j6 -w all 
make: Entering directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_openfortivpn/openfortivpn/work/openfortivpn-1.22.0'
  GEN      doc/openfortivpn.1
  CC       src/openfortivpn-config.o
  CC       src/openfortivpn-http.o
  CC       src/openfortivpn-hdlc.o
  CC       src/openfortivpn-io.o
  GEN      etc/openfortivpn/config
  CC       src/openfortivpn-ipv4.o
  CC       src/openfortivpn-log.o
  CC       src/openfortivpn-tunnel.o
  CC       src/openfortivpn-main.o
src/tunnel.c: In function ‘ssl_connect’:
src/tunnel.c:1100: warning: ‘ENGINE_load_builtin_engines’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:358)
src/tunnel.c:1101: warning: ‘ENGINE_by_id’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:336)
src/tunnel.c:1107: warning: ‘ENGINE_init’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:620)
src/tunnel.c:1110: warning: ‘ENGINE_free’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:493)
src/tunnel.c:1113: warning: ‘ENGINE_set_default_RSA’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:684)
src/tunnel.c:1116: warning: ‘ENGINE_finish’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:628)
src/tunnel.c:1117: warning: ‘ENGINE_free’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:493)
src/tunnel.c:1124: warning: ‘ENGINE_ctrl_cmd’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:452)
src/tunnel.c:1136: warning: ‘ENGINE_load_private_key’ is deprecated (declared at /opt/local/libexec/openssl3/include/openssl/engine.h:639)
  CC       src/openfortivpn-xml.o
  CC       src/openfortivpn-userinput.o
  CCLD     openfortivpn
make: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_openfortivpn/openfortivpn/work/openfortivpn-1.22.0'
```

--- a/configure.ac
+++ b/configure.ac
@@ -166,7 +166,11 @@ write \
 ], [], AC_MSG_ERROR([Required function not found]))
 
 # Checks for optional functions.
-AC_CHECK_FUNCS([pthread_mutexattr_setrobust])
+AC_CHECK_FUNCS([ \
+pthread_mutexattr_setrobust \
+vdprintf \
+])
+
 # Use PKG_CHECK_MODULES compiler/linker flags
 save_openssl_CPPFLAGS="${CPPFLAGS}"
 save_openssl_LIBS="${LIBS}"

--- a/src/userinput.c
+++ b/src/userinput.c
@@ -167,6 +167,22 @@ static int pinentry_read(int from, char **retstr)
 	return -1;
 }
 
+#ifndef HAVE_VDPRINTF
+static int vdprintf(int fd, const char *format, va_list ap)
+{
+	char buffer[2049];
+	int size = vsnprintf(buffer, sizeof(buffer), format, ap);
+
+	if (size < 0)
+		return size;
+
+	if (size >= sizeof(buffer)) // silently discard beyond the buffer size
+		size = sizeof(buffer) - 1;
+
+	return (int) write(fd, buffer, size);
+}
+#endif
+
 static int pinentry_exchange(int to, int from, char **retstr,
                              const char *format, ...)
 {

